In [1]:
import pandas as pd
products_data = pd.read_excel('Data.xlsx', sheet_name='Products')
transactions_data = pd.read_excel('Data.xlsx', sheet_name='Transactions')

In [2]:
products_data.head()
#products_data.columns

Product name  Category  Price
0               JBL Charge 4     Audio   2990
1  Bose QuietComfort Earbuds     Audio   7490
2            Sony WH-1000XM4     Audio  10490
3                LG OLED55CX  Televize  38990
4           Samsung QN55Q80T  Televize  29990

In [3]:
transactions_data.head()
#transactions_data.columns

Transaction ID       Date               Product name  Quantity
0               1 2022-02-01           Samsung QN55Q80T         1
1               1 2022-02-01        Apple iPhone 12 Pro         1
2               2 2022-02-02         Lenovo Tab P11 Pro         2
3               2 2022-02-02         Xiaomi Poco X3 Pro         1
4               3 2022-02-03  Bose QuietComfort Earbuds         2

In [4]:
all_data = transactions_data.merge(products_data, how='left')
all_data

Transaction ID       Date               Product name  Quantity  \
0                 1 2022-02-01           Samsung QN55Q80T         1   
1                 1 2022-02-01        Apple iPhone 12 Pro         1   
2                 2 2022-02-02         Lenovo Tab P11 Pro         2   
3                 2 2022-02-02         Xiaomi Poco X3 Pro         1   
4                 3 2022-02-03  Bose QuietComfort Earbuds         2   
..              ...        ...                        ...       ...   
136              77 2022-05-27                LG 75NANO81         1   
137              77 2022-05-27     Samsung Galaxy Tab S7+         1   
138              78 2022-05-29  Bose QuietComfort Earbuds         2   
139              78 2022-05-29        Apple iPhone 12 Pro         1   
140              78 2022-05-29                LG OLED55CX         1   

             Category  Price  
0            Televize  29990  
1    Mobilní telefony  27990  
2             Tablety  11990  
3    Mobilní telefony   5690  
4               Audio   7490  
..                ...    ...  
136          Televize  44990  
137           Tablety  24990  
138             Audio   7490  
139  Mobilní telefony  27990  
140          Televize  38990  

[141 rows x 6 columns]

## Q1

### Na jaké kategorii produktů máme největší obrat? A zajímalo by mě i jestli se to v jednotlivých měsících mění.

Přidá dva sloupce do 'all_data': 'Revenue' a 'Month'.
Adds two columns to 'all_data': 'Revenue' and 'Month'

In [17]:
all_data['Date'] = pd.to_datetime(all_data['Date'])
all_data['Month'] = all_data['Date'].dt.month
all_data['Revenue'] = (all_data['Quantity'] * all_data['Price']).astype(float)
all_data


Transaction ID       Date               Product name  Quantity  \
0                 1 2022-02-01           Samsung QN55Q80T         1   
1                 1 2022-02-01        Apple iPhone 12 Pro         1   
2                 2 2022-02-02         Lenovo Tab P11 Pro         2   
3                 2 2022-02-02         Xiaomi Poco X3 Pro         1   
4                 3 2022-02-03  Bose QuietComfort Earbuds         2   
..              ...        ...                        ...       ...   
136              77 2022-05-27                LG 75NANO81         1   
137              77 2022-05-27     Samsung Galaxy Tab S7+         1   
138              78 2022-05-29  Bose QuietComfort Earbuds         2   
139              78 2022-05-29        Apple iPhone 12 Pro         1   
140              78 2022-05-29                LG OLED55CX         1   

             Category  Price  Month  Revenue  DayOfWeek  
0            Televize  29990      2  29990.0    Tuesday  
1    Mobilní telefony  27990      2  27990.0    Tuesday  
2             Tablety  11990      2  23980.0  Wednesday  
3    Mobilní telefony   5690      2   5690.0  Wednesday  
4               Audio   7490      2  14980.0   Thursday  
..                ...    ...    ...      ...        ...  
136          Televize  44990      5  44990.0     Friday  
137           Tablety  24990      5  24990.0     Friday  
138             Audio   7490      5  14980.0     Sunday  
139  Mobilní telefony  27990      5  27990.0     Sunday  
140          Televize  38990      5  38990.0     Sunday  

[141 rows x 9 columns]

Definuje, která kategorie má největší obrat

Defines which category has the highest revenue

In [18]:
each_cat_total = all_data.groupby('Category')['Revenue'].sum().reset_index()
each_cat_total.rename(columns={'Revenue': 'Total'}).sort_values(by=['Total'], ascending=False)

Category      Total
3          Televize  1205680.0
2           Tablety   623660.0
1  Mobilní telefony   614460.0
0             Audio   416370.0

Zobrazuje změny pro každý měsíc

Shows changes for each month

In [7]:
max_cat_revenue = all_data.groupby(['Month', 'Category'])['Revenue'].sum().reset_index()
max_cat_revenue_pivot = max_cat_revenue.loc[max_cat_revenue.groupby('Month')['Revenue'].idxmax()]\
    .pivot(
        columns='Month',
        index='Category',
        values='Revenue'
    )
max_cat_revenue_pivot


Month            2         3         4         5
Category                                        
Televize  290920.0  380900.0  266930.0  266930.0

Kategorie "Televize" má největší obrat

Category 'Televize' has the highest revenue

## Q2

### Který den v týdnu je nejsilnější na počet objednávek?

Přidá nový sloupec 'DayOfWeek' a definuje den v týdnu.

Adds new column 'DayOfWeek' and defines the day of the week

In [8]:
all_data['DayOfWeek'] = all_data['Date'].dt.day_name()
all_data

Transaction ID       Date               Product name  Quantity  \
0                 1 2022-02-01           Samsung QN55Q80T         1   
1                 1 2022-02-01        Apple iPhone 12 Pro         1   
2                 2 2022-02-02         Lenovo Tab P11 Pro         2   
3                 2 2022-02-02         Xiaomi Poco X3 Pro         1   
4                 3 2022-02-03  Bose QuietComfort Earbuds         2   
..              ...        ...                        ...       ...   
136              77 2022-05-27                LG 75NANO81         1   
137              77 2022-05-27     Samsung Galaxy Tab S7+         1   
138              78 2022-05-29  Bose QuietComfort Earbuds         2   
139              78 2022-05-29        Apple iPhone 12 Pro         1   
140              78 2022-05-29                LG OLED55CX         1   

             Category  Price  Month  Revenue  DayOfWeek  
0            Televize  29990      2  29990.0    Tuesday  
1    Mobilní telefony  27990      2  27990.0    Tuesday  
2             Tablety  11990      2  23980.0  Wednesday  
3    Mobilní telefony   5690      2   5690.0  Wednesday  
4               Audio   7490      2  14980.0   Thursday  
..                ...    ...    ...      ...        ...  
136          Televize  44990      5  44990.0     Friday  
137           Tablety  24990      5  24990.0     Friday  
138             Audio   7490      5  14980.0     Sunday  
139  Mobilní telefony  27990      5  27990.0     Sunday  
140          Televize  38990      5  38990.0     Sunday  

[141 rows x 9 columns]

Počítá počet produktů v transakci uique a třídí je podle dnů v týdnu.

Counts number of products in the uique transaction and sorts by day of the week

In [9]:
products_in_transaction = all_data.groupby('Transaction ID')['DayOfWeek'].value_counts().reset_index()
products_in_transaction.rename(columns={'count': 'Products'}, inplace=True)
products_in_transaction

Transaction ID  DayOfWeek  Products
0                1    Tuesday         2
1                2  Wednesday         2
2                3   Thursday         1
3                4     Friday         1
4                5   Saturday         2
..             ...        ...       ...
73              74  Wednesday         3
74              75     Sunday         2
75              76    Tuesday         3
76              77     Friday         2
77              78     Sunday         3

[78 rows x 3 columns]

Počítá počet transakcí za den

Counts number of transactions by day

In [10]:
max_transactions_by_days = products_in_transaction.groupby('DayOfWeek')['Transaction ID'].count().reset_index()
max_transactions_by_days.rename(columns={'Transaction ID': 'Transactions'}, inplace=True)
max_transactions_by_days.sort_values(by=['Transactions'], ascending=False)

DayOfWeek  Transactions
1     Monday            13
2   Saturday            13
0     Friday            12
6  Wednesday            11
3     Sunday            10
5    Tuesday            10
4   Thursday             9

Počítá počet produktů za den

Counts number of products by day

In [11]:
max_products_by_day = all_data['DayOfWeek'].value_counts().reset_index()
max_products_by_day.rename(columns={'count': 'products'}, inplace=True)
max_products_by_day

DayOfWeek  products
0   Saturday        29
1     Monday        25
2  Wednesday        22
3     Friday        20
4    Tuesday        16
5   Thursday        15
6     Sunday        14

Většina transakcí a produktů probíhá v sobotu

Most transactions and products are on Saturday

## Q3

### Která kategorie se prodává nejčastěji spolu s produkty z kategorie Televize (resp. jsou spolu v jedné objednávce)? 

In [12]:
transaction_cat = all_data.get(['Transaction ID', 'Category'])
transaction_cat

Transaction ID          Category
0                 1          Televize
1                 1  Mobilní telefony
2                 2           Tablety
3                 2  Mobilní telefony
4                 3             Audio
..              ...               ...
136              77          Televize
137              77           Tablety
138              78             Audio
139              78  Mobilní telefony
140              78          Televize

[141 rows x 2 columns]

Získání ID transakce s kategorií 'Televize' 

Getting transaction id with the category 'Televize' 

In [13]:
transactions_tele = transaction_cat[transaction_cat['Category'].isin(['Televize'])]
ids = transactions_tele['Transaction ID']
print(list(ids))

[1, 5, 8, 9, 13, 19, 21, 25, 31, 33, 37, 38, 42, 45, 46, 50, 51, 53, 55, 55, 58, 61, 62, 65, 68, 70, 72, 72, 74, 76, 77, 78]


In [14]:
transactions_ids = transaction_cat[transaction_cat['Transaction ID'].isin(ids) & (transaction_cat['Category']!='Televize')]
transactions_ids['Category'].value_counts().reset_index()

Category  count
0             Audio     13
1  Mobilní telefony     12
2           Tablety      7

Kategorie "Audio" se nejčastěji prodává společně s produkty z kategorie "Televize".

Category 'Audio' is most often sold together with products from the category 'Televize'

## Q4

### Od 18. 3. 2022 jsem klukům z marketingu navýšil budget na online marketingových platformách (Google Ads, Sklik, Facebook). Dokážeš mi říct, jestli to vedlo k nějaké změně v prodeji?

In [15]:
sales_per_day = all_data.groupby('Date')['Revenue'].sum().reset_index()
sales_per_day

Date  Revenue
0  2022-02-01  57980.0
1  2022-02-02  29670.0
2  2022-02-03  14980.0
3  2022-02-04  25490.0
4  2022-02-05  46480.0
..        ...      ...
64 2022-05-18  71970.0
65 2022-05-22  23970.0
66 2022-05-24  46170.0
67 2022-05-27  69980.0
68 2022-05-29  81960.0

[69 rows x 2 columns]

In [16]:
after_march = sales_per_day.loc[sales_per_day['Date'] >= '2022-03-18']
befor_march = sales_per_day.loc[sales_per_day['Date'] < '2022-03-18']
after_march['Revenue'].sum() - befor_march['Revenue'].sum()

318250.0

Příjmy se zvýšily na 318250 po 18. březnu

Revenue increased on 318250 after March 18th